### DISCUSSION TITLE
First try: Device not recognized _TZE204_myd45weu and empty exposes definitions

### DISCUSSION THREAD

**Discussion by gianarb** [Koenkk/zigbee2mqtt#23094] [Devices]
Thank you for your work. I am new to Zigbee, so I am a bit overwhelmed, and I am not sure what the right expectation is. Most likely, I am doing something wrong, but I don't know what to look for.

I have Home Assistant running with the Mosquitto and the MQTT integration connected. Everything looks up and running, zigbee2mqtt started, and I connected my first device, a soil sensor I bought from AliExpress.

https://it.aliexpress.com/item/1005005377522279.html

```
Jun 18 12:12:08 snowflake zigbee2mqtt[26267]: Zigbee2MQTT:warn  2024-06-18 12:12:08: Received message from unsupported device with Zigbee model 'TS0601' and manufacturer name '_TZE204_myd45weu'
Jun 18 12:12:08 snowflake zigbee2mqtt[26267]: Zigbee2MQTT:warn  2024-06-18 12:12:08: Please see: https://www.zigbee2mqtt.io/advanced/support-new-devices/01_support_new_devices.html
```

![2024-06-18-14-43-00_2549x233](https://github.com/Koenkk/zigbee2mqtt/assets/1630267/af2141aa-a707-4947-96f6-6518f916af10)

It does not expose any value: The `Expose` tab says "Empty exposes definition".

I did the process as explained in the tutorial.

When I try to send commands via Dev Console I get this error:

`error:         z2m: EventBus error 'Publish/mqttMessage': Cannot read properties of null (reading 'toZigbee')`

HA does not recognize it as a device and looking at Mosquitto I only see the `zigbee2mqtt` topic. I am not sure if I am right but if I understood correctly in order to see a device in home assistant it should write to a topic prefixed with `homeassistant` by default. This is probably a consequence of the device timing out.

This is the ZigBee hub I am using, and I didn't flash a firmware, I am using the stock one:

https://www.amazon.it/dp/B0B6P22YJC?psc=1

Since I am new to all of this I am not sure about what to do next 

Mosquitto 2.0.18
Zigbee2MQTT 1.38.0

**Comment by gianarb** [comment]
I bet it is something related to all of this typescript over here, but I am not sure about who to validate that:

https://github.com/Slaw3km/zigbee-herdsman-converters/blob/c56a3781c83fe2d47dd337da918a149e6c8fa9df/src/devices/tuya.ts

**Comment by gianarb** [comment]
Ok I figured out a good conversion so far, I am gonna keep digging because now I figured out the full cycle and if I can get to something useful I will submit the converstion to herdsman

**Comment by gianarb** [comment]
I also figured out that I need to update herdsman since the work for this sensor is all done and released

https://github.com/Koenkk/zigbee-herdsman-converters/pull/7593/files

**Reply by jdukat** [reply]
I have the same problem - see my post on Home Assistant forum:
https://community.home-assistant.io/t/tuya-soil-sensor-not-recognized-by-zigbee2mqtt/741851/9

Did you figure out how to update converters in Home Assistant?

**Comment by gianarb** [comment]
I am not at home now so I am not sure how precise I can be. But it is not a problem with home assistant in my case but with the version of ZigBee2mqtt that depends on a conversion of the converter that does not have this device. At the moment btw I copied the Js conversation from the converter repository into my server and I linked it to ZigBee2mqtt as described by the article I got from the logs.

As soon as I get the next ZigBee2mqtt version from my Linux distro I will remove it.

Hope it helps

**Reply by jdukat** [reply]
I understand that it's the zigbeee2mqtt that requires an update, which is not part of the latest release 1.38.0.
Did you get it working according to [these instructions](https://www.zigbee2mqtt.io/advanced/support-new-devices/02_support_new_tuya_devices.html#_4-mapping-the-datapoints)?
I tried to do it but with no success.
I ended up switching to Zigbee2MQTT Edge in my Home Assistant and now the device works just fine. I'll switch back to the stable version after the new release.

**Comment by gianarb** [comment]
Great I think what you do just works.
No i ended up coping the JS code that converts this device from the main repository to my insurance and linking it to the my ZigBee2mqtt configuration like described here

<https://www.zigbee2mqtt.io/advanced/support-new-devices/01_support_new_devices.html>

It was easy option for my current situation.

**Reply by jdukat** [reply]
Ok, I don't know why it didn't work for me because I also tried the same approach, must have done something wrong, but problem solved anyway :) Thanks

**Reply by Schoermen87** [reply]
> Großartig, ich denke, was Sie tun, funktioniert einfach. Nein, ich habe am Ende den JS-Code kopiert, der dieses Gerät vom Hauptrepository in meine Versicherung konvertiert, und es mit meiner ZigBee2mqtt-Konfiguration verknüpft, wie hier beschrieben < https://www.zigbee2mqtt.io/advanced/support-new-devices/01_support_new_devices.html >. Es war eine einfache Option für meine aktuelle Situation.
> […](#)
> Am 22. Juni 2024 um 19:57:51 Uhr GMT+02:00 schrieb „Jarosław Dukat“ ***@***.***>: Ich verstehe, dass zigbeee2mqtt ein Update benötigt, das nicht Teil der neuesten Version 1.38.0 ist. Haben Sie es gemäß [dieser Anleitung]( https://www.zigbee2mqtt.io/advanced/support-new-devices/02_support_new_tuya_devices.html#_4-mapping-the-datapoints ) zum Laufen gebracht? Ich habe es versucht, aber ohne Erfolg. Schließlich bin ich in meinem Home Assistant auf Zigbee2MQTT Edge umgestiegen und jetzt funktioniert das Gerät einwandfrei. Nach der neuen Version werde ich wieder auf die stabile Version umsteigen. – Antworten Sie direkt auf diese E-Mail oder sehen Sie sie sich auf GitHub an: [#23094 (Antwort im Thread)](https://github.com/Koenkk/zigbee2mqtt/discussions/23094#discussioncomment-9848204) Sie erhalten dies, weil Sie den Öffnungs-/Schließstatus geändert haben. Nachrichten-ID: ***@***.***>

Können Sie den JS-Code hier zeigen? Bei mir geht es auch nicht

**Comment by Cic0r** [comment]
Same problem for me, won't work. :(
Device is not supported, but fingerprint is available, says GitHub...
Tried it with a external converter, too, but still not supported. Is there a mistake maybe?

```
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const modernExtend = require('zigbee-herdsman-converters/lib/modernExtend');
const e = exposes.presets;
const ea = exposes.access;
const tuya = require('zigbee-herdsman-converters/lib/tuya');

const definition = {
    // Since a lot of Tuya devices use the same modelID, but use different datapoints
    // it's necessary to provide a fingerprint instead of a zigbeeModel
    fingerprint: [
        {
            // The model ID from: Device with modelID 'TS0601' is not supported
            // You may need to add \u0000 at the end of the name in some cases
            modelID: 'TS0601_soil',
            // The manufacturer name from: Device with modelID 'TS0601' is not supported.
            manufacturerName: '_TZE204_myd45weu',
        },
    ],
    model: 'TS0601_new',
    vendor: 'Tuya',
    description: 'Soil sensor',
    fromZigbee: [tuya.fz.datapoints],
    toZigbee: [tuya.tz.datapoints],
    onEvent: tuya.onEventSetTime, // Add this if you are getting no converter for 'commandMcuSyncTime'
    configure: tuya.configureMagicPacket,
    exposes: [e.temperature(), e.soil_moisture(), tuya.exposes.temperatureUnit(), e.battery(), tuya.exposes.batteryState()],
    meta: {
        // All datapoints go in here
        tuyaDatapoints: [
                [3, 'soil_moisture', tuya.valueConverter.raw],
                [5, 'temperature', tuya.valueConverter.raw],
                [9, 'temperature_unit', tuya.valueConverter.temperatureUnit],
                [14, 'battery_state', tuya.valueConverter.batteryState],
                [15, 'battery', tuya.valueConverter.raw],
        ],
    },
    extend: [
        // A preferred new way of extending functionality.
    ],
};

module.exports = definition;
```

**Reply by phillipsnick** [reply]
@Cic0r I got it working with the following on zigbee2mqtt v1.38.0

Small workaround until zigbee2mqtt starts using zigbee-herdsman-convertersv19.49.0

Hope this helps :) 

```js
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const tuya = require('zigbee-herdsman-converters/lib/tuya');

const e = exposes.presets;
const definition = {
    fingerprint: tuya.fingerprint('TS0601', ['_TZE200_myd45weu', '_TZE200_ga1maeof', '_TZE200_2se8efxh', '_TZE204_myd45weu']),
    model: 'TS0601_soil',
    vendor: 'Tuya',
    description: 'Soil sensor',
    fromZigbee: [tuya.fz.datapoints],
    toZigbee: [tuya.tz.datapoints],
    configure: tuya.configureMagicPacket,
    exposes: [e.temperature(), e.soil_moisture(), tuya.exposes.temperatureUnit(), e.battery(), tuya.exposes.batteryState()],
    meta: {
        tuyaDatapoints: [
            [3, 'soil_moisture', tuya.valueConverter.raw],
            [5, 'temperature', tuya.valueConverter.raw],
            [9, 'temperature_unit', tuya.valueConverter.temperatureUnit],
            [14, 'battery_state', tuya.valueConverter.batteryState],
            [15, 'battery', tuya.valueConverter.raw],
        ],
    },
}

module.exports = definition;
```

**Reply by KiCADHugo** [reply]
Thank you, now ist works.

**Reply by Cic0r** [reply]
> @Cic0rIch habe es mit dem folgenden auf zigbee2mqtt v1.38.0 zum Laufen gebracht
> 
> Kleiner Workaround, bis zigbee2mqtt zigbee-herdsman-convertersv19.49.0 verwendet
> 
> Hoffe das hilft :)
> 
> ```js
> const exposes = require('zigbee-herdsman-converters/lib/exposes');
> const tuya = require('zigbee-herdsman-converters/lib/tuya');
> 
> const e = exposes.presets;
> const definition = {
>     fingerprint: tuya.fingerprint('TS0601', ['_TZE200_myd45weu', '_TZE200_ga1maeof', '_TZE200_2se8efxh', '_TZE204_myd45weu']),
>     model: 'TS0601_soil',
>     vendor: 'Tuya',
>     description: 'Soil sensor',
>     fromZigbee: [tuya.fz.datapoints],
>     toZigbee: [tuya.tz.datapoints],
>     configure: tuya.configureMagicPacket,
>     exposes: [e.temperature(), e.soil_moisture(), tuya.exposes.temperatureUnit(), e.battery(), tuya.exposes.batteryState()],
>     meta: {
>         tuyaDatapoints: [
>             [3, 'soil_moisture', tuya.valueConverter.raw],
>             [5, 'temperature', tuya.valueConverter.raw],
>             [9, 'temperature_unit', tuya.valueConverter.temperatureUnit],
>             [14, 'battery_state', tuya.valueConverter.batteryState],
>             [15, 'battery', tuya.valueConverter.raw],
>         ],
>     },
> }
> 
> module.exports = definition;
> ```

Thank you, but still Not working for me. :(

Edit:  Updated Z2M to 1.39.0 and it is working without extra .js. :)

**Reply by KiCADHugo** [reply]
How have you updated to 1.39.0? I use npm install zigbee2mqtt and it is 1.38.0

**Reply by Cic0r** [reply]
> Wie haben Sie auf 1.39.0 aktualisiert? Ich verwende npm install zigbee2mqtt und es ist 1.38.0

There was an Update in HA, No manuell Update.
![Screenshot_2024-07-02-16-41-57-344_io homeassistant companion android-edit](https://github.com/Koenkk/zigbee2mqtt/assets/95544216/921182bb-7cc1-4483-8e8a-f3971968a8e7)
